pandas/core/arrays/integer.py

```diff
@@ -481,7 +482,16 @@ def astype(self, dtype, copy: bool = True) -> ArrayLike:
     def _values_for_factorize(self) -> Tuple[np.ndarray, float]:
         # TODO: https://github.com/pandas-dev/pandas/issues/30037
         # use masked algorithms, rather than object-dtype / np.nan.
-        return self.to_numpy(na_value=np.nan), np.nan
+        return self.to_numpy(dtype=float, na_value=np.nan), np.nan
+
+    def factorize2(self, na_sentinel: int = -1) -> Tuple[np.ndarray, "ExtensionArray"]:
```

Comment:
This is of course not meant to stay. I think we have two options:

- Override the base class `factorize` here on IntegerArray to use the mask (so basically rename `factorize2` to `factorize`)
- Expand the `_values_for_factorize` spec to allow it to return a mask instead of na_value. The base class `factorize` then would need to choose the correct path depending on whether `_values_for_factorize` returned a boolean ndarray, or something else (scalar na_value)

Short term, the first is the easiest. But long term, I think the second would be nice to allow external EAs to more easily use this as well (avoiding the need they have to override the base class `factorize` as well). But this second option has some more implications (need to update everywhere it is used internally).

Reply:
I also think that expanding `_values_for_factorize` is the way to go, but I worry about inferring the EA wants a masked based on `na_value` being a boolean ndarray. It's an edge case, but consider a nested / ragged array where each "scalar" element is an ndarray.

Could we instead (or also) have a class attribute like `_masked_factorize = True` indicating that they want to used the masked variant?

Reply:
> It's an edge case, but consider a nested / ragged array where each "scalar" element is an ndarray.

Even in such a case, I would expect the na_value not to be an array. But OK, it's certainly not the most robust way.

An alternative could also be to return 3 values (values, None, mask) in case of a mask, and then we can check the number of items returned. 
Also not the cleanest solution, though. I just also don't really like the `_masked_factorize` attribute .. Certainly if we would start doing this in other places as well, and then might need other similar attributes. Unless we would go for a single `_is_masked` attribute and then you need to go all or nothing.

Reply:
option 2 (return a mask) would be my preference, even if its a breaking change, much cleaner

Reply:
I don't think we should simply break this (we would break other projects like GeoPandas), since it is relatively easily avoidable. 

If we want to replace `na_value` in factorize with passing a mask (with a given `na_value` you can compute the mask in advance with `values == na_value`), we could deprecate the option of returning an `na_value` from `_values_for_factorize`, though.